samples/JavaFxPlayVideoAndAudio.java

```diff
@@ -56,28 +86,65 @@ public void start(final Stage primaryStage) throws Exception {
+
+                    final DataLine.Info info = new DataLine.Info(SourceDataLine.class, audioFormat);
+                    soundLine = (SourceDataLine) AudioSystem.getLine(info);
+                    soundLine.open(audioFormat);
+                    soundLine.start();
+                    playbackTimer = new PlaybackTimer(soundLine);
+                } else {
+                    soundLine = null;
+                    playbackTimer = new PlaybackTimer();
+                }
 
-                final DataLine.Info info = new DataLine.Info(SourceDataLine.class, audioFormat);
-                final SourceDataLine soundLine = (SourceDataLine) AudioSystem.getLine(info);
-                soundLine.open(audioFormat);
-                soundLine.start();
+                final JavaFXFrameConverter converter = new JavaFXFrameConverter();
 
-                final Java2DFrameConverter converter = new Java2DFrameConverter();
+                final ExecutorService audioExecutor = Executors.newSingleThreadExecutor();
+                final ExecutorService imageExecutor = Executors.newSingleThreadExecutor();
 
-                ExecutorService executor = Executors.newSingleThreadExecutor();
+                final long maxReadAheadBufferMicros = 1000 * 1000L;
 
+                long lastTimeStamp = -1L;
                 while (!Thread.interrupted()) {
-                    Frame frame = grabber.grab();
+                    final Frame frame = grabber.grab();
                     if (frame == null) {
                         break;
                     }
+                    if (lastTimeStamp < 0) {
+                        playbackTimer.start();
+                    }
+                    lastTimeStamp = frame.timestamp;
                     if (frame.image != null) {
-                        final Image image = SwingFXUtils.toFXImage(converter.convert(frame), null);
-                        Platform.runLater(new Runnable() { public void run() {
-                            imageView.setImage(image);
-                        }});
+                        final Frame imageFrame = frame.clone();
+
+                        imageExecutor.submit(new Runnable() {
+                            public void run() {
+                                final Image image = converter.convert(imageFrame);
+                                final long timeStampDeltaMicros = imageFrame.timestamp - playbackTimer.elapsedMicros();
+                                imageFrame.close();
+                                if (timeStampDeltaMicros > 0) {
+                                    final long delayMillis = timeStampDeltaMicros / 1000L;
+                                    try {
+                                        Thread.sleep(delayMillis);
+                                    } catch (InterruptedException e) {
+                                        e.printStackTrace();
```

Comment:
Let's just put the typical `Thread.currentThread().interrupt()` here?

Reply:
done